evm/bridge-token-factory/contracts/BridgeTokenFactory.sol

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.24;
+
+import {AccessControlUpgradeable} from "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import {ERC1967Proxy} from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+import {UUPSUpgradeable} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+import "./BridgeToken.sol";
+import "./SelectivePausableUpgradable.sol";
+import "./Borsh.sol";
+
+contract BridgeTokenFactory is
+    UUPSUpgradeable,
+    AccessControlUpgradeable,
+    SelectivePausableUpgradable
+{
+    enum WhitelistMode {
+        NotInitialized,
+        Blocked,
+        CheckToken,
+        CheckAccountAndToken
+    }
+
+    // We removed ProofConsumer from the list of parent contracts and added this gap
+    // to preserve storage layout when upgrading to the new contract version.
+    uint256[54] private __gap;
+
+    mapping(address => string) private _ethToNearToken;
+    mapping(string => address) private _nearToEthToken;
+    mapping(address => bool) private _isBridgeToken;
+
+    mapping(string => WhitelistMode) private _whitelistedTokens;
+    mapping(bytes => bool) private _whitelistedAccounts;
+    bool private _isWhitelistModeEnabled;
+
+    address public tokenImplementationAddress;
+    address public nearBridgeDerivedAddress;
+
+    mapping(uint128 => bool) public completedTransfers;
+
+    bytes32 public constant PAUSABLE_ADMIN_ROLE = keccak256("PAUSABLE_ADMIN_ROLE");
+    uint constant UNPAUSED_ALL = 0;
+    uint constant PAUSED_WITHDRAW = 1 << 0;
+    uint constant PAUSED_DEPOSIT = 1 << 1;
+
+    struct BridgeDeposit {
+        uint128 nonce;
+        string token;
+        uint128 amount;
+        address recipient;
+        string feeRecipient;
+    }
+
+    struct MetadataPayload {
+        string token;
+        string name;
+        string symbol;
+        uint8 decimals;
+    }
+
+    // Event when funds are withdrawn from Ethereum back to NEAR.
+    event Withdraw(
+        string token,
+        address indexed sender,
+        uint256 amount,
+        string recipient,
+        address indexed tokenEthAddress
+    );
+
+    event Deposit(
+        string indexed token,
+        uint256 amount,
+        address recipient,
+        uint128 indexed nonce,
+        string indexed feeRecipient
```

Comment:
I indexed `feeRecipient` because an off-chain service may want to search for events related to a specific recipient (to collect all the fees, for example). Yes, it will have to hash the account Id before performing the search.

`token` can be useful for analytics. `recipient`, I agree, is the least useful

Reply:
Removed indexing of `feeRecipient` as hashing would complicate the logic when claiming fees on Near.

Still not sure about `token` and `recipient`. I view events not only from POV of utility that allows us to execute some logic on Near, but also as a source of additional information about user activity. This way it makes sense to keep these properties.